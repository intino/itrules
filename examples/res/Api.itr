rule type(api, service)
	public class $name+firstUpperCase+SnakeCaseToCamelCase~Service {
		ModelSupplier supplier;

		public void setup(int port) {
			Javalin.create()
				<<$resource...[$NL]>>
				.start(port)
		}

		<<$schema...[$NL$NL]>>
	}

rule trigger(resource)
    <<.$operation...[$NL.]>>

rule trigger(operation)
	$method+lowerCase("<<$<root>basePath>><<$path>>", ctx -> <<$response(>>new $<container>name+ClassName~Controller(ctx, supplier).$method+lowerCase($parameter...[, ])<<$response+closeMethod>>))

rule trigger(schema)
	public record $name($attribute+constructor...[, ]) {
		public void check() {
			try {
			<<if ($attribute+checkNotNull...[ || ]) throw new io.javalin.http.BadRequestResponse("Malformed BookInput parameter");>>
			<<$attribute+checkFormat>>
			} catch (Exception e) {
				throw new io.javalin.http.BadRequestResponse("Malformed BookInput parameter");
			}
		}
	}

rule type(attribute) and trigger(constructor)
	$type $name+firstLowerCase

rule type(attribute) and type(required) and trigger(checkNotNull)
	$name() == null

rule type(attribute) and attribute(type, date) and trigger(checkFormat)
	java.text.DateFormat.getInstance().parse($name());

rule (attribute(in, path) or attribute(in, query)) and type(required) and trigger(parameter)
	ctx.$in~ParamAsClass("$name", $type.class).getOrThrow(e -> new io.javalin.http.BadRequestResponse("$name parameter not found"))

rule (attribute(in,path) or attribute(in, query)) and trigger(parameter)
	ctx.$in~ParamAsClass("$name", $type+ClassName.class)

rule attribute(in, body) and trigger(parameter)
	check(ctx.bodyAsClass($type+ClassName.class))

rule trigger(closeMethod)
	)