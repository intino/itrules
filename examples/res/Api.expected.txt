public class Service {
	ModelSupplier supplier;

	public void setup(int port) {
		Javalin.create()
			.get("/api/books", ctx ->new booksController(ctx, supplier).get()))
			.post("/api/books", ctx ->new booksController(ctx, supplier).post()))
			.get("/api/book", ctx ->new booksController(ctx, supplier).get(ctx.pathParamAsClass("id", integer.class).getOrThrow(e -> new io.javalin.http.BadRequestResponse("id parameter not found")))))
			.put("/api/book", ctx ->new booksController(ctx, supplier).put(ctx.pathParamAsClass("id", integer.class).getOrThrow(e -> new io.javalin.http.BadRequestResponse("id parameter not found")), check(ctx.bodyAsClass(BookInput.class)))))
			.start(port)
	}

	public record BookInput(String title, Integer authorId, java.util.Date publicationDate, String genre) {
		public void check() {
			try {
			if (title() == null || authorId() == null) throw new io.javalin.http.BadRequestResponse("Malformed BookInput parameter");
			} catch (Exception e) {
				throw new io.javalin.http.BadRequestResponse("Malformed BookInput parameter");
			}
		}
	}

	public record Book(Integer id, String title, Integer authorId, java.util.Date publicationDate, String genre) {
		public void check() {
			try {
			if (id() == null || title() == null || authorId() == null || publicationDate() == null || genre() == null) throw new io.javalin.http.BadRequestResponse("Malformed BookInput parameter");
			} catch (Exception e) {
				throw new io.javalin.http.BadRequestResponse("Malformed BookInput parameter");
			}
		}
	}
}